### DISCUSSION TITLE
Is there a way to commit migrations before querying indexes?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#11834] [Q&A]
I have a module that is called "Teams". This module is responsible of providing team structure, in addition to way to set current team for all user.

This module register the following migration 

```csharp
public class TeamMigrations : DataMigration
{
    private readonly IContentDefinitionManager _contentDefinitionManager;

    public TeamMigrations(IContentDefinitionManager contentDefinitionManager) => _contentDefinitionManager = contentDefinitionManager;

    public int Create()
    {
        _contentDefinitionManager.AlterPartDefinition("TeamPart", part => part
            .Attachable()
            .WithField("Team", field => field
                .OfType(nameof(ContentPickerField))
                .WithDisplayName("Team")
                .MergeSettings<ContentPickerFieldSettings>(settings =>
                {
                    settings.Required = true;
                    settings.DisplayedContentTypes = new[] { "Team" };
                })
            )
        );

        _contentDefinitionManager.AlterPartDefinition("ParentTeamPart", part => part
            .Attachable()
            .WithDisplayName("Parent Team Part")
            .WithField("Team", field => field
                .OfType(nameof(ContentPickerField))
                .WithDisplayName("Parent Team")
                .MergeSettings<ContentPickerFieldSettings>(settings =>
                {
                    settings.DisplayedContentTypes = new[] { "Team" };
                })
            )
        );

        SchemaBuilder.CreateMapIndexTable<TeamIndex>(table => table
           .Column<string>("DisplayText", column => column.WithLength(255))
           .Column<string>("ContentItemId", column => column.WithLength(26))
        );

        _contentDefinitionManager.AlterTypeDefinition("Team", type => type
                .Listable()
                .Creatable()
                .DisplayedAs("Teams")
                .WithPart("TitlePart")
                .WithPart("ParentTeamPart", part => part
                    .WithDescription("Parent Team")
                )
            );

        return 1;
    }
}
```

As you can see, this migration create an index, and updated content definition.

**Important** This module requires that every user has a default team as all content items filtered based on teams you belong to/have access to. So if you belong to team **[A]**, then you'll see content for users that belong to team **[A]** and any sub teams of **[A]**. For this to work property, I have to set the current team for every user when this module is enabled. The logic that I should execute is as follow

 1. Check the `TeamIndex` table to see if there is record already exists, and if so pick the first one.
 2. If no team exists, create a new default team.
 3. Pull all the users from the database, and foreach user attach `UserTeamPart` part by setting the `TeamId` property to the default teamId.

I thought implementing the `IFeatureEventHandler` would be a perfect place for this logic since I can hook into these events every time the module is installed or enabled and process my logic to ensure that every user has a team. The code will look something like this

```csharp
private async Task UpdateUsersAsync(IFeatureInfo feature)
      {
          if (feature.Id != Startup.ModuleName)
          {
              return;
          }

           _contentManager??= _serviceProvider.GetRequiredService<IContentManager>();

            var team = await _session.Query<ContentItem, TeamIndex>().FirstOrDefaultAsync();

            if (team == null)
            {
                // if this is the first time we enable this module, we'll create a default team.
               // if the setup recipe creates default teams, then we should use what the recipe created instead of creating a non-needed/redundant default team.
                team = await _contentManager.NewAsync("Team");
                await _contentManager.CreateAsync("Team");
            }

          var users = await _session.Query<User, UserIndex>().ListAsync();

          foreach (var user in users)
          {
              // if the user already has the UserTeamPart attached, no don't see to do anything
              if (user.Has<UserTeamPart>())
              {
                  continue;
              }

              user.Put(new UserTeamPart()
              {
                  TeamId = team.ContentItemId,
              });

              _session.Save(user);
          }
      }
```

The issue with the code above is, during site-setup, the code is executed in the same sql-transaction as the migration. So the call ` _session.Query<User, UserIndex>` will throw an exception as the `UserIndex` table won't exists since the sql-transaction is not yet committed.

Additionally, since my setup recipe will create default team, I have to execute the above code **after** the migrations are executed, the recipe finished executing and the code is committed to the database.

### Question
How can I ensure the my code above is executed every time a feature is enabled/installed, yet after the migrations are processed? 

I thought about calling `_session.SaveChangesAsync()` just before this call `_session.Query<ContentItem, TeamIndex>().FirstOrDefaultAsync()` but that did not work. I thought calling `_session.SaveChangesAsync()` would first commit all the changes to the database and close the transaction and then process the next set of queries using another transaction. Apparently that isn't what happens here.

**Comment by ns8482e** [comment]
I guess you use DeferredTask when there's schema change
